transport/peer.go

```diff
@@ -0,0 +1,60 @@
+package transport
+
+//go:generate mockgen -destination=transporttest/peer.go -package=transporttest go.uber.org/yarpc/transport PeerIdentifier,Peer,SubscribablePeer,PeerSubscriber
+
+// PeerStatus maintains information about the Peer's connection state
+type PeerStatus string
+
+const (
+	// PeerAvailable indicates the Peer is available for requests
+	PeerAvailable PeerStatus = "available"
+	// PeerConnecting indicates the Peer is in the process of connecting
+	PeerConnecting PeerStatus = "connecting"
+	// PeerUnavailable indicates the Peer is unavailable for requests
+	PeerUnavailable PeerStatus = "unavailable"
+)
+
+/*
+PeerIdentifier is able to uniquely identify a peer (e.g. hostport)
+*/
+type PeerIdentifier interface {
+	Identifier() string
+}
+
+/*
+Peer is a level on top of PeerIdentifier.  It should be created by a PeerAgent so we
+can maintain multiple references to the same downstream peer (e.g. hostport).  This is
+useful for load balancing requests to downstream services.
+*/
+type Peer interface {
+	PeerIdentifier
+
+	GetStatus() PeerStatus // Get the status of the Peer
+	GetAgent() PeerAgent   // Get the PeerAgent which owns this Peer
+	Pending() int          // The number of pending requests on this peer
+	IncPending()           // Increment the number of pending requests
+	DecPending()           // Decrement the number of pending requests
+}
+
+/*
+SubscribablePeer (Name Pending) is the "Full" interface for a Peer, it should only be used by the PeerAgent.
+It adds functions for adding/removing PeerSubscriber references.  If a Peer has no subscriptions/references
+then it should be safe to delete.
+*/
+type SubscribablePeer interface {
+	Peer
+
+	OnRetain(PeerSubscriber) error  // Tell Peer it is being tracked by PeerSubscriber
+	OnRelease(PeerSubscriber) error // Tell Peer it is no longer being tracked by PeerSubscriber
```

Comment:
What's the advantage of embedding peer here rather than making it an argument
to the functions?

```
type PeerSubscription interface {
  OnRetain(Peer, PeerSubscriber) error
  OnRelease(Peer, PeerSubscriber) error
  References(Peer) int 
}
```

Specializing interfaces with embedding is dangerous. It doesn't play well with
wrapping. That is, if you define,

```
func wrapPeer(p Peer) Peer {
    return myPeer{p}
}

type myPeer struct{ Peer }
```

You lose information that a peer is subscribable when you wrap it.

```
_, ok := p.(SubscribablePeer)
fmt.Println(ok)  // true

p2 := wrapPeer(p)
_, ok = p2.(SubscribablePeer)
fmt.Println(ok)  // false
```